pkg/ingressCache/safeTrie.go

```diff
@@ -0,0 +1,168 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing
+permissions and limitations under the License.
+
+In addition, you may not use the software for any purposes that are
+illegal under applicable law, and the grant of the foregoing license
+under the Apache 2.0 license is conditioned upon your compliance with
+such restriction.
+*/
+
+package ingresscache
+
+import (
+	"slices"
+	"sync"
+
+	"github.com/dghubble/trie"
+	"github.com/nuclio/errors"
+)
+
+type SafeTrie struct {
+	pathTrie *trie.PathTrie
+	rwMutex  sync.RWMutex
+}
+
+// NewSafeTrie creates a new SafeTrie instance
+func NewSafeTrie() *SafeTrie {
+	return &SafeTrie{
+		pathTrie: trie.NewPathTrie(),
+		rwMutex:  sync.RWMutex{},
+	}
+}
+
+// SetFunctionName sets a function for a given path. If the path does not exist, it creates it
+func (st *SafeTrie) SetFunctionName(path string, function string) error {
+	if path == "" {
+		return errors.New("path is empty")
+	}
+
+	if function == "" {
+		return errors.New("function is empty")
+	}
+
+	st.rwMutex.Lock()
+	defer st.rwMutex.Unlock()
+
+	// get the exact path value in order to avoid creating a new path if it already exists
+	pathValue := st.pathTrie.Get(path)
+	if pathValue == nil {
+		st.pathTrie.Put(path, []string{function})
+		return nil
+	}
+
+	pathFunctionNames, ok := pathValue.([]string)
+	if !ok {
+		return errors.Errorf("value is not a []string, got: %T", pathValue)
+	}
+
+	if slices.Contains(pathFunctionNames, function) {
+		// If the function already exists at this path, skip adding it to prevent duplicates
+		return nil
+	}
+
+	pathFunctionNames = append(pathFunctionNames, function)
+	st.pathTrie.Put(path, pathFunctionNames)
+
+	return nil
+}
+
+// DeleteFunctionName removes a function from a path and also deletes the path if the function is the only one associated with that path
+func (st *SafeTrie) DeleteFunctionName(path string, function string) error {
+	st.rwMutex.Lock()
+	defer st.rwMutex.Unlock()
+
+	pathValue := st.pathTrie.Get(path)
+	if pathValue == nil {
+		// If pathValue is nil, the path does not exist, so nothing to delete
+		return nil
+	}
+
+	pathFunctionNames, ok := pathValue.([]string)
+	if !ok {
+		return errors.Errorf("path value should be []string, got %T", pathValue)
+	}
+
+	// If the function to delete matches the current function name and it's the only value, delete the path
+	if len(pathFunctionNames) == 1 {
+		if pathFunctionNames[0] == function {
+			st.pathTrie.Delete(path)
+		}
+		return nil
+	}
+
+	// TODO - will be removed once moving into efficient pathFunctionNames implementation (i.e. not using slices)
+	pathFunctionNames = excludeElemFromSlice(pathFunctionNames, function)
+	st.pathTrie.Put(path, pathFunctionNames)
+	return nil
+}
+
+// GetFunctionName retrieve the closest prefix matching the path and returns the associated functions
+func (st *SafeTrie) GetFunctionName(path string) ([]string, error) {
```

Comment:
you return the slice, so:
```suggestion
// GetFunctionNames retrieve the closest prefix matching the path and returns the associated functions
func (st *SafeTrie) GetFunctionNames(path string) ([]string, error) {
```

Reply:
Done here - [CR comment - rename directory and interfaces comments](https://github.com/v3io/scaler/pull/70/commits/e18346adc0bc000f548851fa7d0a589dcca8c7b3)